examples/command_bot.rs

```diff
@@ -0,0 +1,123 @@
+use std::sync::Arc;
+use std::time::{SystemTime, UNIX_EPOCH};
+use std::{env, process::exit};
+
+use js_int::UInt;
+use matrix_sdk::{
+    self,
+    events::room::message::{MessageEvent, MessageEventContent, TextMessageEventContent},
+    AsyncClient, AsyncClientConfig, EventEmitter, Room, SyncSettings,
+};
+use tokio::sync::{Mutex, RwLock};
+use url::Url;
+
+struct CommandBot {
+    /// This clone of the `AsyncClient` will send requests to the server,
+    /// while the other keeps us in sync with the server using `sync_forever`.
+    client: Mutex<AsyncClient>,
+    /// A timestamp so we only respond to messages sent after the bot is running.
+    start_time: UInt,
+}
+
+impl CommandBot {
+    pub fn new(client: AsyncClient) -> Self {
+        let now = SystemTime::now();
+        let timestamp = now.duration_since(UNIX_EPOCH).unwrap().as_millis();
+        Self {
+            client: Mutex::new(client),
+            start_time: UInt::new(timestamp as u64).unwrap(),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl EventEmitter for CommandBot {
+    async fn on_room_message(&self, room: Arc<RwLock<Room>>, event: &MessageEvent) {
+        let (msg_body, timestamp) = if let MessageEvent {
+            content: MessageEventContent::Text(TextMessageEventContent { body: msg_body, .. }),
+            origin_server_ts,
+            ..
+        } = event
+        {
+            (msg_body.clone(), *origin_server_ts)
+        } else {
+            (String::new(), UInt::min_value())
+        };
+
+        if msg_body.contains("!party") && timestamp > self.start_time {
```

Comment:
Another way to do this is to do an initial sync without the event emitter registered, register the emitter and only then do the `sync_forever()` call. The timestamps usually can be faked so it's probably a bit better to not rely on them.

Doesn't need to be changed for this to get merged.